src/csharp/Grpc.Tools/GeneratorServices.cs

```diff
@@ -61,71 +62,114 @@ protected bool GrpcOutputPossible(ITaskItem proto)
+                    filename = Path.Combine(baseNamespaceDir, filename);
+                }
+            }
+
             outputs[0] = Path.Combine(outdir, filename) + ".cs";
 
             if (doGrpc)
             {
                 // Override outdir if kGrpcOutputDir present, default to proto output.
                 string grpcdir = protoItem.GetMetadata(Metadata.GrpcOutputDir);
-                filename = LowerUnderscoreToUpperCamelGrpcWay(basename);
-                outputs[1] = Path.Combine(
-                    grpcdir != "" ? grpcdir : outdir, filename) + "Grpc.cs";
+                outdir = grpcdir != "" ? grpcdir : outdir;
+                outputs[1] = Path.Combine(outdir, filename) + "Grpc.cs";
             }
             return outputs;
         }
 
-        // This is how the gRPC codegen currently construct its output filename.
-        // See src/compiler/generator_helpers.h:118.
-        string LowerUnderscoreToUpperCamelGrpcWay(string str)
+        string GetNamespaceDir(string filename, string baseNamespace)
         {
-            var result = new StringBuilder(str.Length, str.Length);
-            bool cap = true;
-            foreach (char c in str)
-            {
-                if (c == '_')
-                {
-                    cap = true;
-                }
-                else if (cap)
-                {
-                    result.Append(char.ToUpperInvariant(c));
-                    cap = false;
+            if (!File.Exists(filename)) return null;
+
+            string ns = GetFileNamespace(filename);
+            if (ns == null) return null;
+
+            string namespace_suffix = ns;
+            if (!string.IsNullOrEmpty(baseNamespace)) {
+                // Check that the base_namespace is either equal to or a leading part of
+                // the file namespace. This isn't just a simple prefix; "Foo.B" shouldn't
+                // be regarded as a prefix of "Foo.Bar". The simplest option is to add "."
+                // to both.
+                string extended_ns = ns + ".";
+                if (extended_ns.IndexOf(baseNamespace + ".", StringComparison.Ordinal) != 0) {
+                    return null; // This will be ignored, because we've set an error.
                 }
-                else
-                {
-                    result.Append(c);
+
+                namespace_suffix = ns.Substring(baseNamespace.Length);
+                if (namespace_suffix.IndexOf(".", StringComparison.Ordinal) == 0) {
+                    namespace_suffix = namespace_suffix.Substring(1);
                 }
             }
-            return result.ToString();
+
+            return namespace_suffix.Replace('.', Path.DirectorySeparatorChar);
+        }
+
+        private string GetFileNamespace(string filename)
+        {
+            string data = File.ReadAllText(filename);
+
+            // First try to match the namespace from file options.
+            Match match = s_namespaceRegex.Match(data);
```

Comment:
matching the regexes like this is very fragile and dirty. We cannot do it like this.
A simple counterexample:  if I include `option  csharp_namespace` (with two spaces) instead of `option csharp_namespace` the regex won't catch that and you'll silently generate the wrong path and things will break in a very hard to detect way. This problem is fundamental and you'll never be able to come up with a regex that catches all the interesting cases reliably.

Basically the issue is that you're trying to emulate the c# namespace mangling logic from the plugin's C++ code, (without having access to any of the protobuf primitives that could tell you values of some of the options you need) and that is simply not going to work.